src/loaders/xm_load.c

```diff
@@ -944,6 +945,73 @@ static int xm_load(struct module_data *m, HIO_HANDLE * f, const int start)
 		}
 	}
 
+#ifndef LIBXMP_CORE_PLAYER
+	/* Load MPT properties from the end of the file. */
+	while (1) {
+		uint32 ext = hio_read32b(f);
+		uint32 sz = hio_read32l(f);
+		int known = 0;
+
+		if (hio_error(f) || sz > LONG_MAX)
+			break;
+
+		switch (ext) {
+		case MAGIC4('t','e','x','t'):		/* Song comment */
+			known = 1;
+			if (m->comment != NULL)
+				break;
+
+			if ((m->comment = (char *)malloc(sz + 1)) == NULL)
+				break;
+
+			sz = hio_read(m->comment, 1, sz, f);
+			m->comment[sz] = '\0';
+
+			for (i = 0; i < (int)sz; i++) {
+				int b = m->comment[i];
+				if (b == '\r') {
+					m->comment[i] = '\n';
+				} else if ((b < 32 || b > 127) && b != '\n'
+					   && b != '\t') {
+					m->comment[i] = '.';
+				}
+			}
+			break;
+
+		case MAGIC4('M','I','D','I'):		/* MIDI config */
+		case MAGIC4('P','N','A','M'):		/* Pattern names */
+		case MAGIC4('C','N','A','M'):		/* Channel names */
+		case MAGIC4('C','H','F','X'):		/* Channel plugins */
+		case MAGIC4('X','T','P','M'):		/* Inst. extensions */
+			known = 1;
+			/* fall-through */
+
+		default:
+			/* Plugin definition */
+			if ((ext & MAGIC4('F','X', 0, 0)) == MAGIC4('F','X', 0, 0))
+				known = 1;
+
+			if (sz) hio_seek(f, sz, SEEK_CUR);
+			break;
+		}
+
+		if(known && claims_ft2)
+			is_mpt_116 = 1;
+
+		if (ext == MAGIC4('X','T','P','M'))
```

Comment:
Note that it would be good if this also exited the loop when encountering `STPM` in addition to `XTPM`... right now if no instrument extensions are used, OpenMPT still writes an `XTPM` tag with `STPM` following immediately afterwards (because I was expecting this kind of code which relies on both being always present), but I might want to change this in the future.

Reply:
OK, noted!